packages/turbo-workspaces/src/managers/npm.ts

```diff
@@ -0,0 +1,188 @@
+  1. package-lock.json
+  2.
+*/
+async function detect(args: DetectArgs): Promise<boolean> {
+  const lockFile = path.join(args.workspaceRoot, "package-lock.json");
+  const packageManager = getWorkspacePackageManager({
+    workspaceRoot: args.workspaceRoot,
+  });
+  return fs.existsSync(lockFile) || packageManager === "npm";
+}
+
+/*
+ Read workspace data from npm workspaces into generic format
+*/
+async function read(args: ReadArgs): Promise<Project> {
+  const isNpm = await detect(args);
+  if (!isNpm) {
+    throw new ConvertError("Not an npm workspaces project");
+  }
+
+  const packageJson = getPackageJson(args);
+  return {
+    name: getWorkspaceName(args),
+    packageManager: "npm",
+    paths: {
+      root: args.workspaceRoot,
+      packageJson: path.join(args.workspaceRoot, "package.json"),
+      lockfile: path.join(args.workspaceRoot, "package-lock.json"),
+      nodeModules: path.join(args.workspaceRoot, "node_modules"),
+    },
+    workspaceData: {
+      globs: packageJson.workspaces || [],
+      workspaces: expandWorkspaces({
+        workspaceGlobs: packageJson.workspaces,
+        ...args,
+      }),
+    },
+  };
+}
+
+/*
+ Create npm workspaces from generic format
+ Creating npm workspaces involves:
+
+  1. Adding the workspaces field in package.json
+  2. Setting the packageManager field in package.json
+  3. Updating all workspace package.json dependencies to ensure correct format
+*/
+async function create(args: CreateArgs): Promise<void> {
+  const { project, options, to, logger } = args;
+
+  logger.mainStep(`Creating npm workspaces`);
+  const packageJson = getPackageJson({ workspaceRoot: project.paths.root });
+  logger.rootHeader();
+  if (project.packageManager !== "yarn") {
+    logger.rootStep(
+      `adding "workspaces" field to ${project.name} root "package.json"`
+    );
+    packageJson.workspaces = project.workspaceData.globs;
+  }
+  logger.rootStep(
+    `adding "packageManager" field to ${project.name} root "package.json"`
+  );
+  packageJson.packageManager = `${to.name}@${to.version}`;
+
+  if (!options?.dry) {
+    fs.writeJSONSync(project.paths.packageJson, packageJson, { spaces: 2 });
+  }
+
+  // if we're converting from pnpm, we need to update the workspace package.json files
+  if (project.packageManager === "pnpm") {
+    // root dependencies
+    updateDependencies({
+      workspace: { name: "root", paths: project.paths },
+      project,
+      to,
+      logger,
+      options,
+    });
+
+    logger.workspaceHeader();
+    project.workspaceData.workspaces.forEach((workspace) =>
+      updateDependencies({ workspace, project, to, logger, options })
+    );
+  }
+}
+
+/*
+Remove npm workspace data
+
+Removing npm workspaces involves:
+  1. Removing the workspaces field from package.json
+*/
+async function remove(args: RemoveArgs): Promise<void> {
+  const { project, options, to, logger } = args;
+
+  logger.mainStep(`Removing npm workspaces`);
+  if (to.name !== "yarn") {
```

Comment:
How about we _always_ remove, and then _always_ create?

I think that'll be required for dealing with the bonus weird `npm` format anyway.

Reply:
I did think about that, but my thinking here was that I wanted to try and touch as little as possible on the way from a to b. 

However, it does make these functions a little more gross since each package manager needs to care about more than just itself. 

I'll play with this and see how it feels.